versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractWriteUpdateCommits.java

```diff
@@ -0,0 +1,122 @@
+import org.projectnessie.versioned.persist.adapter.CommitLogEntry;
+import org.projectnessie.versioned.persist.adapter.DatabaseAdapter;
+import org.projectnessie.versioned.persist.adapter.ImmutableCommitLogEntry;
+
+public abstract class AbstractWriteUpdateCommits {
+
+  private final DatabaseAdapter databaseAdapter;
+  private final boolean commitWritesValidated;
+
+  protected AbstractWriteUpdateCommits(
+      DatabaseAdapter databaseAdapter, boolean commitWritesValidated) {
+    this.databaseAdapter = databaseAdapter;
+    this.commitWritesValidated = commitWritesValidated;
+  }
+
+  Hash intAsHash(int i) {
+    return Hash.of(String.format("%08x", i));
+  }
+
+  private CommitLogEntry helloCommit(int i) {
+    return CommitLogEntry.of(
+        i,
+        intAsHash(i),
+        i,
+        singletonList(intAsHash(i - 1)),
+        ByteString.copyFromUtf8("hello " + i),
+        emptyList(),
+        emptyList(),
+        0,
+        null,
+        emptyList(),
+        emptyList(),
+        emptyList());
+  }
+
+  @Test
+  void writeMultipleCommits() throws Exception {
+    List<Hash> hashes =
+        IntStream.rangeClosed(1, 10).mapToObj(this::intAsHash).collect(Collectors.toList());
+    List<CommitLogEntry> commits =
+        hashes.stream()
+            .mapToInt(h -> Integer.parseInt(h.asString(), 16))
+            .mapToObj(this::helloCommit)
+            .collect(Collectors.toList());
+
+    databaseAdapter.writeMultipleCommits(commits);
+
+    try (Stream<CommitLogEntry> retrieved =
+        databaseAdapter.fetchCommitLogEntries(hashes.stream())) {
+      assertThat(retrieved).containsExactlyInAnyOrderElementsOf(commits);
+    }
+
+    List<CommitLogEntry> updatedCommits =
+        commits.stream()
+            .map(
+                commit ->
+                    ImmutableCommitLogEntry.builder()
+                        .from(commit)
+                        .metadata(ByteString.copyFromUtf8("updated " + commit.getHash().asString()))
+                        .build())
+            .collect(Collectors.toList());
+
+    databaseAdapter.updateMultipleCommits(updatedCommits);
+
+    try (Stream<CommitLogEntry> retrieved =
+        databaseAdapter.fetchCommitLogEntries(hashes.stream())) {
+      assertThat(retrieved).containsExactlyInAnyOrderElementsOf(updatedCommits);
+    }
+  }
+
+  @Test
+  void validateWriteUpdateCommitPreconditions() throws Exception {
+    assumeThat(commitWritesValidated).isTrue();
+
+    List<Hash> hashes =
+        IntStream.rangeClosed(1, 10).mapToObj(this::intAsHash).collect(Collectors.toList());
+    List<CommitLogEntry> commits =
+        hashes.stream()
+            .mapToInt(h -> Integer.parseInt(h.asString(), 16))
+            .mapToObj(this::helloCommit)
+            .collect(Collectors.toList());
+
+    databaseAdapter.writeMultipleCommits(commits.subList(0, 2));
+    assertThatThrownBy(() -> databaseAdapter.writeMultipleCommits(commits.subList(0, 2)))
```

Comment:
are we intentionally only writing the sublist?
would an exception still be thrown if we wrote the first 3 elements? i.e. it included 1 non existing ?
if this behavior has certain limitations that we need to respect in this test, then maybe explain them in a comment here.